packages/versioning/src/validate.ts

```diff
@@ -0,0 +1,413 @@
+              target: namespace,
+            });
+          }
+        } else {
+          if (typeof value !== "string") {
+            reportDiagnostic(program, {
+              code: "versioned-dependency-not-string",
+              format: { dependency: program.checker.getNamespaceString(dependencyNs) },
+              target: namespace,
+            });
+          }
+        }
+      }
+    },
+  });
+  validateVersionedNamespaceUsage(program, namespaceDependencies);
+}
+
+function validateVersionedNamespaceUsage(
+  program: Program,
+  namespaceDependencies: Map<NamespaceType | undefined, Set<NamespaceType>>
+) {
+  for (const [source, targets] of namespaceDependencies.entries()) {
+    const dependencies = source && getVersionDependencies(program, source);
+    for (const target of targets) {
+      const targetVersions = getVersion(program, target);
+
+      if (targetVersions !== undefined && dependencies?.get(target) === undefined) {
+        reportDiagnostic(program, {
+          code: "using-versioned-library",
+          format: {
+            sourceNs: program.checker.getNamespaceString(source),
+            targetNs: program.checker.getNamespaceString(target),
+          },
+          target: source ?? NoTarget,
+        });
+      }
+    }
+  }
+}
+
+interface IncompatibleVersionValidateOptions {
+  isTargetADependent?: boolean;
+}
+
+/**
+ * Validate the target versioning is compatible with the versioning of the soruce.
+ * e.g. The target cannot be added after the source was added.
+ * @param source Source type referencing the target type.
+ * @param target Type being referenced from the source
+ */
+function validateTargetVersionCompatible(
+  program: Program,
+  source: Type,
+  target: Type,
+  validateOptions: IncompatibleVersionValidateOptions = {}
+) {
+  let targetVersionRange = getResolvedVersionRange(program, target);
+  if (targetVersionRange === undefined) {
+    return;
+  }
+
+  const sourceVersionRange = getResolvedVersionRange(program, source);
+
+  const [sourceNamespace, sourceVersions] = getVersions(program, source);
+  const [targetNamespace, targetVersions] = getVersions(program, target);
+  if (sourceNamespace === undefined || sourceVersions === undefined) {
+    return;
+  }
+  if (targetNamespace === undefined || targetNamespace === undefined) {
+    return;
+  }
+
+  if (sourceNamespace !== targetNamespace) {
+    const versionMap = getVersionDependencies(program, (source as any).namespace)?.get(
+      targetNamespace
+    );
+    if (versionMap === undefined) {
+      return;
+    }
+    targetVersionRange = translateVersionRange(
+      program,
+      targetVersionRange,
+      versionMap,
+      targetVersions!,
+      source,
+      target
+    );
+    if (targetVersionRange === undefined) {
+      return;
+    }
+  }
+
+  if (sourceVersionRange === undefined) {
+    if (!validateOptions.isTargetADependent) {
```

Comment:
Hmm, if I am reading this correctly, we are depending on the source range coming from `@added` and `@removed` decorators placed explicitly on a property or its enclosing type, but all types inherently have an addedOn matching the lowest version of their namespace.

let's say that a service namespace foo at version '1' took a versioned dependency on a namespace  bar at version A, so  foo:1 -> bar: A

then

```
@versioned("a" | "b")
namespace Bar {
  @added("b")
  model BarSpecificType {
  ...
  }
}
@versioned("1" | "2")
@versionedDependency(Bar, {"1": "b"; "2": "b"})
namespace foo {
...
model Baz {
  myProperty: Bar.BarSpecificType
}
```

I believe that property Baz.myProperty would have an undefined sourceVersionRange, and so would throw here, even though the type is compatible.  I may be missing something about how ranges are set up, however, but certainly testing implicit addedOn would be important

Reply:
The degenerate case is that we should allow versioned dependencies for unversioned namespaces - that is, namespaces that have no explicit version (single service version) should be able to take a versioned dependency on another namespace that has versions.